ext/regex.go

```diff
@@ -0,0 +1,357 @@
+	}
+	return re, nil
+}
+
+// Initializing regular expression patterns for validating replacement strings.
+var (
+	reGroupNum     = regexp.MustCompile(`\$(\d+)`)
+	reGroupName    = regexp.MustCompile(`\$\{([a-zA-Z_][a-zA-Z0-9_]*)\}`)
+	reGroupInvalid = regexp.MustCompile(`\$[a-zA-Z_][a-zA-Z0-9_]*`)
+)
+
+func validateReplacementString(re *regexp.Regexp, replaceStr string) error {
+	// If there are no $N or ${name} patterns, skip validation.
+	if !strings.Contains(replaceStr, "$") {
+		return nil
+	}
+
+	groupNames := re.SubexpNames()
+	groupCount := len(groupNames) - 1 // Exclude group 0 (whole match)
+	// Find all $N patterns in the replacement string and validate them.
+	matches := reGroupNum.FindAllStringSubmatch(replaceStr, -1)
+	for _, m := range matches {
+		if len(m) > 1 {
+			idx, _ := strconv.Atoi(m[1])
+			if idx < 0 || idx > groupCount {
+				return fmt.Errorf("replacement string references group $%d, but regex has only %d group(s)", idx, groupCount)
+			}
+		}
+	}
+
+	if strings.Contains(replaceStr, "${") {
+		validNames := make(map[string]struct{})
+		for _, name := range groupNames {
+			if name != "" {
+				validNames[name] = struct{}{}
+			}
+		}
+		// If there are named groups, validate them against the defined group names.
+		nameMatches := reGroupName.FindAllStringSubmatch(replaceStr, -1)
+		for _, m := range nameMatches {
+			if len(m) > 1 {
+				if _, ok := validNames[m[1]]; !ok {
+					return fmt.Errorf("invalid capture group name in replacement string: %s", m[1])
+				}
+			}
+		}
+	}
+
+	// Check for invalid $word references (e.g., $a)
+	invalids := reGroupInvalid.FindAllString(replaceStr, -1)
+	for _, m := range invalids {
+		// If not matched by $N, it's invalid
+		if !reGroupNum.MatchString(m) {
+			return fmt.Errorf("invalid group reference: %s", m)
+		}
+	}
+
+	return nil
+}
+
+func captureFirstMatch(target, regexStr ref.Val) ref.Val {
+	t := string(target.(types.String))
+	r := string(regexStr.(types.String))
+	re, err := compileRegex(r)
+	if err != nil {
+		return types.WrapErr(err)
+	}
+
+	matches := re.FindStringSubmatch(t)
+	if len(matches) == 0 {
+		return types.OptionalNone
+	}
+
+	// If there is a capturing group, return the first group; otherwise, return the whole match.
+	if len(matches) > 1 {
+		capturedGroup := matches[1]
+		// If optional group is empty, return OptionalNone.
+		if capturedGroup == "" {
+			return types.OptionalNone
+		}
+		return types.OptionalOf(types.String(matches[1]))
+	}
+	return types.OptionalOf(types.String(matches[0]))
+}
+
+func captureAllMatches(target, regexStr ref.Val) ref.Val {
+	t := string(target.(types.String))
+	r := string(regexStr.(types.String))
+	re, err := compileRegex(r)
+	if err != nil {
+		return types.WrapErr(err)
+	}
+
+	matches := re.FindAllStringSubmatch(t, -1)
+	var result []string
+	if len(matches) == 0 {
+		return types.NewStringList(types.DefaultTypeAdapter, result)
+	}
+
+	hasCaptureGroups := len(matches[0]) > 1
```

Comment:
So if there are no groups, this just returns a list of all non-overlapping matches?

What happens when there are overlapping capture groups? For `captureAll('abcdef', r'((abc)def)')` is it `['abc', 'abcdef']`?

Separately, it might be better to unnest this a little bit for readability. e.g.

```
if len(match[0] == 1) {
  // no capture groups
  for _, match := range matches {
    result = append(result, match[0]
  }
  return types.NewStringList(...)
}

// handle capture groups
for _, match := range matches {
  for i, submatch := range match {
    if i == 0 {
      // skip overall match
      continue;
     }
    if (submatch != "") {
      result = append(result, submatch)
    }
  }
  result = append(result, match[0]
}
```